src/Crypto/KmsMaterialsProvider.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace Aws\Crypto;
+
+use Aws\Kms\KmsClient;
+
+/**
+ * Uses KMS to supply materials for encrypting and decrypting data.
+ */
+class KmsMaterialsProvider extends MaterialsProvider
+{
+    private $kmsClient;
+    private $kmsKeyId;
+
+    /**
+     * @param KmsClient $kmsClient A KMS Client for use encrypting and
+     *                             decrypting keys.
+     * @param string $kmsKeyId The private KMS key id to be used for encrypting
+     *                         and decrypting keys.
+     */
+    public function __construct(
+        KmsClient $kmsClient,
+        $kmsKeyId = null
+    ) {
+        $this->kmsClient = $kmsClient;
+        $this->kmsKeyId = $kmsKeyId;
+    }
+
+    public function fromDecryptionEnvelope(MetadataEnvelope $envelope)
+    {
+        if (empty($envelope[MetadataEnvelope::MATERIALS_DESCRIPTION_HEADER])) {
+            throw new \RuntimeException('Not able to detect kms_cmk_id from an'
+                . ' empty materials description.');
+        }
+
+        $materialsDescription = json_decode(
+            $envelope[MetadataEnvelope::MATERIALS_DESCRIPTION_HEADER],
+            true
+        );
+        if (empty($materialsDescription['kms_cmk_id'])) {
+            throw new \RuntimeException('Not able to detect kms_cmk_id from kms'
+                . ' materials description.');
+        }
+
+        return new KmsMaterialsProvider(
+            $this->kmsClient,
+            $materialsDescription['kms_cmk_id']
```

Comment:
What's a `cmk`? Could abbreviations be avoided in this parameter name, or is the expanded version too long to reasonably type?

Reply:
`'kms_cmk_id'` is how the key is indexed in the materials description, this is [consistent with the Java (and other) SDK implementations](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/KMSEncryptionMaterials.java#L33). The ['KeyId' naming on `Encrypt`](https://github.com/aws/aws-sdk-php/pull/1395/files#diff-46de444f818130423cc34de9445afd90R83) is used for the `$kmsKeyId` variable.